# MDAnderson_Su23
Analysis Scripts Developed During Summer 2023 Research Internship

**comut.py**<br>
- comut plot of the gbm samples

**gene_expression.py**<br>
- plotting the gene expression level across gbm samples based on whether the gene is associated with a neoloop

**genes_in_neoloop.py**<br>
- algorithm for detecting genes that overlap or flank a neoloop

**genes_in_neoloop_vis.py**<br>
- visualization script for the top 20 genes that are recurrently associated with a neoloop

**idh_neoloop.py**<br>
- analyzing the number of neoloops detected in idh mutant samples and idh wild-type samples

**neoloopfinder.py**<br>
- visualizing complex assemblies and neoloops with hic data

**primary_recurrent.py**<br>
- visualizing the structural variant and new loop counts in primary versus recurrent gbm samples

Subtyping information: https://gdc.cancer.gov/about-data/publications/gbm_exp

**subtype_analysis.py**<br>
- classifying gbm subtypes based on gene expression analysis

**subtype_dimreduction.py**<br>
- dimensionality reduction analysis on gene expression profile across gbm samples

**subtype_heatmap.py**<br>
- heatmap of gene expression profile across gbm samples

**subtype_ternary.py**<br>
- ternary plot of gbm subtype classification results

**sv_analysis.py**<br>
- bar plot of the composition of structural variants detected in each gbm samples through wgs and hic
